### PR TITLE
dfu: boot: mcuboot_shell: Prevent erasing MCUboot/app flash areas

### LINKED ISSUE #60443: mcuboot: shell: erase: Allows deleting current image and breaking device
**Describe the bug**
The mcuboot shell command have an erase option, this allows erasing of any flash image number, including that of the currently running image, which should not be allowed.

**Expected behavior**
An error should be emitted if the user attempts to erase the slot of the currently running image

**Impact**
Annoyance

**Environment (please complete the following information):**
 - Commit SHA or Version used: ea2aac6e5131bf916abd1e37334e94cc3481e1a8

### PR DESCRIPTION
**Pull request by nordicjm**
Prevents the user from potentially bricking a device by erasing
    the MCUboot or currently running application flash areas with the
    mcuboot erase command

Fixes #60443

### CODE REVIEW

**Review by de-nordic — DISMISSED** [review]

File: subsys/dfu/boot/mcuboot_shell.c [hunk 1]
```diff
@@ -91,6 +92,22 @@ static int cmd_mcuboot_erase(const struct shell *sh, size_t argc,
...
+#if FIXED_PARTITION_EXISTS(boot_partition)
+	    || id == FIXED_PARTITION_ID(boot_partition)
+#endif
+#if DT_FIXED_PARTITION_EXISTS(DT_CHOSEN(zephyr_code_partition))
+	    || id == DT_FIXED_PARTITION_ID(DT_CHOSEN(zephyr_code_partition))
+#endif
+	) {
+		shell_error(sh, "Cannot erase boot or active partitions");
+		return 0;
+	}
```
Comment (fabiobaltieri) [inline subsys/dfu/boot/mcuboot_shell.c:108]: that's a bit rough imo :-) how about having two separate conditions?

```c
        /* Check if this is the parent (MCUboot) or own slot and if so, deny the request */
#if FIXED_PARTITION_EXISTS(boot_partition)
        if (id == FIXED_PARTITION_ID(boot_partition)) {
                shell_error(sh, "Cannot erase boot partition");
                return 0;
        }
#endif
#if DT_FIXED_PARTITION_EXISTS(DT_CHOSEN(zephyr_code_partition))
        if (id == DT_FIXED_PARTITION_ID(DT_CHOSEN(zephyr_code_partition))) {
                shell_error(sh, "Cannot erase active partitions");
                return 0;
        }
#endif
```

also is `return 0` correct?

Comment (nordicjm) [reply subsys/dfu/boot/mcuboot_shell.c]: Have updated

Comment (fabiobaltieri) [reply subsys/dfu/boot/mcuboot_shell.c]: Thanks

**Review by fabiobaltieri — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Comment by TirelessDev** [comment]
Hey @nordicjm, sorry to dig up an old PR, but this appears to break in situations when partition manager is in play because the dts specified partitions do not match those in the flash map which is used to perform the erase. As it occurs in my own project the external flash area turns out to be area 1 while the internal active partition is actually area 3. This clashes with the defines generated from the device tree which place the active partition as area 1 and thus the erase is blocked. 

Are there any tricks I am missing that might address this mismatch?

**Comment by nordicjm** [comment]
> Hey @nordicjm, sorry to dig up an old PR, but this appears to break in situations when partition manager is in play because the dts specified partitions do not match those in the flash map which is used to perform the erase. As it occurs in my own project the external flash area turns out to be area 1 while the internal active partition is actually area 3. This clashes with the defines generated from the device tree which place the active partition as area 1 and thus the erase is blocked.
> 
> Are there any tricks I am missing that might address this mismatch?

https://github.com/nrfconnect/sdk-zephyr/pull/2114 should fix the issues you are seeing

**Comment by TirelessDev** [comment]
Legend, cheers!

**Pull request merged**